csrc/cpu/cpu_attn_impl.hpp

```diff
@@ -0,0 +1,1984 @@
+
+    void apply_softcap_and_scale(logits_buffer_t* __restrict__ logits_buffer,
+                                 const int64_t logits_buffer_stride,
+                                 int32_t q_head_num, int32_t kv_tile_token_num,
+                                 float softcap_scale, float scale) {
+#ifdef __AVX512F__
+      DEFINE_AVX512_EXP
+#endif
+      float inv_softcap_scale = 1.0 / softcap_scale;
+      vec_op::FP32Vec16 scale_vec(scale);
+      vec_op::FP32Vec16 softcap_scale_vec(softcap_scale);
+      vec_op::FP32Vec16 inv_softcap_scale_vec(inv_softcap_scale);
+      vec_op::FP32Vec16 ones_vec(1.0);
+      logits_buffer_t* __restrict__ curr_logits_buffer = logits_buffer;
+      const int32_t vec_num = kv_tile_token_num / 16;
+      for (int32_t i = 0; i < q_head_num; ++i) {
+        logits_buffer_t* __restrict__ curr_logits_buffer_iter =
+            curr_logits_buffer;
+        for (int32_t j = 0; j < vec_num; ++j) {
+          vec_op::FP32Vec16 vec(curr_logits_buffer_iter);
+          vec = vec * scale_vec * inv_softcap_scale_vec;
+
+#ifdef __AVX512F__
+          vec = avx512_exp_u20(vec);
+          vec_op::FP32Vec16 inv_vec = ones_vec / vec;
+          vec = (vec - inv_vec) / (vec + inv_vec);
+#else
+          vec.save(curr_logits_buffer_iter);
+          for (int k = 0; k < 16; ++k) {
+            curr_logits_buffer_iter[k] = std::tanh(curr_logits_buffer_iter[k]);
+          }
+          vec = vec_op::FP32Vec16(curr_logits_buffer_iter);
+#endif
+          vec = vec * softcap_scale_vec;
+          vec.save(curr_logits_buffer_iter);
+
+          curr_logits_buffer_iter += 16;
+        }
+
+        curr_logits_buffer += logits_buffer_stride;
+      }
+    }
+
+    void apply_alibi_slopes_and_scale(
+        logits_buffer_t* __restrict__ logits_buffer,
+        const float* __restrict__ alibi_slopes,
+        const int64_t logits_buffer_stride, const int32_t q_tile_start_pos,
+        const int32_t kv_tile_start_pos, const int32_t q_token_num,
+        const int32_t kv_tile_token_num, const int32_t q_heads_per_kv,
+        float scale) {
+      alignas(64) constexpr float initial_arange_vals[16] = {
+          0.0f, 1.0f, 2.0f,  3.0f,  4.0f,  5.0f,  6.0f,  7.0f,
+          8.0f, 9.0f, 10.0f, 11.0f, 12.0f, 13.0f, 14.0f, 15.0f};
+      const int32_t vec_num = kv_tile_token_num / 16;
+
+      vec_op::FP32Vec16 initial_arange_vals_vec(initial_arange_vals);
+      initial_arange_vals_vec =
+          initial_arange_vals_vec + vec_op::FP32Vec16((float)kv_tile_start_pos);
+      vec_op::FP32Vec16 scale_vec(scale);
+      vec_op::FP32Vec16 pos_offset_vec(16.0);
+      logits_buffer_t* __restrict__ curr_logits_buffer = logits_buffer;
+      for (int32_t i = 0; i < q_token_num; ++i) {
+        vec_op::FP32Vec16 curr_q_pos_vec((float)(i + q_tile_start_pos));
+        for (int32_t j = 0; j < q_heads_per_kv; ++j) {
+          vec_op::FP32Vec16 alibi_scale_vec(alibi_slopes[j]);
+          vec_op::FP32Vec16 curr_kv_pos_vec = initial_arange_vals_vec;
```

Comment:
This doesn't compile on Arm because all `vec_op::FP32Vec16` constructors are `explicit`
Changing it to: `vec_op::FP32Vec16 curr_kv_pos_vec(initial_arange_vals_vec)` fixes the problem.